### ISSUE TITLE
Mac OS: PaintDC not destroyed correctly

### ISSUE THREAD

**Issue by DietmarSchwertberger** [wxWidgets/Phoenix#680]
**Operating system**: Mac OS Sierra
**wxPython version**:  4.0.0.b2
**Stock or custom build**: Stock
**Python version**: 3.6.1
**Stock or custom build**: Stock

**Description of the problem**:

Currently the PaintDC needs to be explicitely destroyed at the end of a Paint event.

Recently there was a bug report about repaint problems on Mac OS with an embedded matplotlib WXAgg canvas:
https://github.com/matplotlib/matplotlib/issues/10174#issuecomment-357068436

If you have matplotlib 2.1.1 installed, you can run the script that you can find at the top of the report.
It's working fine on Windows. On Mac OS, there are problems on resizing the window. The canvas and other elements are sometimes drawn at wrong positions.
Sometimes, there are reports ` ...Python[413] <Error>: CGContextRestoreGState: invalid context 0x0. Backtrace:`.
A full backtrace is at https://github.com/matplotlib/matplotlib/files/1624146/backtrace.txt
At the top of it you can see things like `dealloc` and `release_wxPaintDC`.

It seems that garbage collection of a previous `PaintDC` does interfere with a current `EVT_PAINT`.
Sometimes, it seems that the C++ context might have been destroyed already when the Python object is collected.

Disabling the garbage collector during the Paint event fixes the problem.
Alternatively an explicit call `drawDC.Destroy()` at the end of `matplotlib.backends.backend_wx.FigureCanvasWx._onPaint` also solves the problem.

An explicit call to `Destroy` usually should not be necessary...

**Comment by RobinD42** [comment]
I just lightly browsed the mpl issue comments so far, but a couple questions come to mind right away. 

* Is it possible that more than one `PaintDC` is being created in the same `EVT_PAINT` handler? 
* What about a `ClientDC`[1] being used while a `PaintDC` is still active? 
* Are DCs being saved and reused later instead of letting them die and just creating a new one as needed? 
* Is there a `self.Update()` happening anywhere that could cause a recursive call to the `EVT_PAINT` handler?

[1] IMO, `ClientDC` should be completely avoided these days. They do not work quite the same on OSX as the other platforms, but changing state and then calling `Refresh` or `Update` to trigger the drawing in the paint handler does work very well everywhere.

**Comment by DietmarSchwertberger** [comment]
I have just checked. Everything looks reasonable and I could not find any of the issues that you listed. Especially, some print calls did confirm that there were no recursive Paint events.

I found one strange thing, though, by inserting some print calls.
After creation `drawDC=PaintDC(self)`, `sys.getrefcount(drawDC)` returns 2, as expected.
The paint handler calls `draw(drawDC)` where the refcount is 4.
`draw` in turn calls `FigureCanvasAgg.draw(self)`. There's no reference to the DC passed to this.
At the end of `draw`, the refcount is still 4.
But after `draw` returns, in the Paint handler the refcount now prints as 3, while it should be 2 again.
I can reproduce this: if `FigureCanvasAgg.draw` is not called or if I disable the garbage collector during this call or if I avoid the recursive mathtext parser then the refcount is 2 at the end of the Paint handler.
The strangest thing is that this occurs already during the very first `EVT_PAINT` where there was no previous instance of `PaintDC`.

**Comment by DietmarSchwertberger** [comment]
OK, I've found a way to avoid the refcount increase by creating it only after the call to `FigureCanvasAgg.draw`, but the repaint problems are remaining.
The refcount increase might be due to some exception handling inside the call which stored a reference to a the stack.

**Comment by DietmarSchwertberger** [comment]
@RobinD42 : The Matplotlib people are starting to prepare the 2.2 release. Do you see a risk with adding `drawDC.Destroy()` to the end of the `EVT_PAINT` handler? Either only for Mac OS or for all.

**Comment by RobinD42** [comment]
I wish the reason it's needed could be figured out, but I don't think it will hurt anything to add it there. 

Also, I just noticed that `FigureCanvasWx._onPaint` currently has `evt.Skip()` at the end. That should be removed unless there are other `EVT_PAINT` bindings for that window that still need to be called, as skipping paint events on MS Windows can lead to infinite immediate refire of the paint event.

**Comment by DietmarSchwertberger** [comment]
The `evt.Skip()` is out already, as I found that it caused redraw problems on Windows ( https://github.com/matplotlib/matplotlib/pull/10192 ) . Sorry, I should have written this above as it's not obvious from the linked Matplotlib issue.
I will keep an eye on the Matplotlib release planning and ensure that the `Destroy` call will be in 2.2 if no other solution is found before.

Does wxPython rely on the refcount going to 0 when the event handler returns? The above example makes clear that this is not necessarily the case.

**Comment by RobinD42** [comment]
> The evt.Skip() is out already,

Oops, apparently I didn't update to current master before I looked in the code.

> Does wxPython rely on the refcount going to 0 when the event handler returns?

Yes. The normal life cycle of wrapped classes is that the C++ object is created when the Python proxy object is created and the proxy object is given ownership of the C++ object, which means that the C++ object will be deleted when the proxy object is cleaned up. We automatically transfer ownership back to C++ in a number of places (like when creating a child widget as the C++ object will be owned and managed by the parent widget.) In those cases nothing extra is done when the proxy object is cleaned up.

`wx.Object.Destroy` bypasses some of that as it tells the object to destroy itself and also transfers ownership away from Python so the proxy object will not try to destroy the C++ object again.

**Comment by DietmarSchwertberger** [comment]
I'm wondering whether this is reliable & reproducible enough. If a C++ `PaintDC` object is on the stack, I think the destructor is guaranteed to be called immediately when the handler returns. But with Python garbage collection? I think there's no guarantee for an immediate call, even though it should occur usually if there are no cyclic references.

I don't understand this as with a `sys.getrefcount(drawDC)` of 2 at the end of the handler, I thought it would be 0 after return and the destructor called immediately. But still the effects were there. With the garbage collector disabled during repaint they were gone.

Do the Python proxies created by SIP behave differently? Are they collected later?

**Comment by RobinD42** [comment]
> Do the Python proxies created by SIP behave differently? 

No, they should be basically the same as any other object type.

**Comment by mesalu** [comment]
> I don't understand this as with a sys.getrefcount(drawDC) of 2 at the end of the handler, I thought it would be 0

There are two references at this point. One being `drawDC` and the other being the reference on the callstack for `sys.getrefcount`. 

How frequent are these paint events? Would it be reasonably possible for the GC to not have had time to do a collection in between? 

Its possible that the PaintDC wrapper object isn't playing well with the GC, or that its just not getting collected in time. 

If you nest the paint handler in a function that calls `gc.collect` afterwards does it resolve the problem?

**Comment by DietmarSchwertberger** [comment]
Well, my understanding of the Python documentation was that a destructor would be called immediately when the ref count reaches 0.
In the unpatched matplotlib, there seems to be one additional ref count, maybe due to a handled exception.
So there it is understandable that the object is only collected later and wx can not rely on the destructor being called at the end of the paint handler.
As the draw routine is quite memory intensive due to recursion, the gc runs during the next paint event at least sometimes.

With a modified paint handler, I can avoid the additional reference, such that it will decrease to 0 right after the event handler is done. But still the painting errors occur.
I can avoid these by explicitely calling the garbage collector at the beginning of the handler or by disabling it temporarily or by calling the `drawDC.Destroy()` myself at the end of the handler.
So it seems that the destructor is being called later.

I think that wxPython should not rely on the garbage collector calling the destructor. There should be a mechanism in the event dispatcher to call `Destroy` on the `PaintDC` that was created during while handling the event.

From the CPython documentation:
- Do not depend on immediate finalization of objects when they become unreachable (so you should always close files explicitly).
- CPython currently uses a reference-counting scheme with (optional) delayed detection of cyclically linked garbage, which collects most objects as soon as they become unreachable, but is not guaranteed to collect garbage containing circular references.

**Comment by mesalu** [comment]
@RobinD42 and I are pretty sure there's no reasonable way to do this in the Py/CPP binding layer with out creating future headaches.

Right now we're toying with the idea of tweaking in python context managers. Use would then look like:

```python
def OnPaint(self):
    with wx.PaintDC(self) as dc:
         # do work....
     # dc is properly destroyed.
```

Using the context manager is optional, and as such will not affect existing code.

**Comment by DietmarSchwertberger** [comment]
I was thinking about some hack like a global variable `_last_created_PaintDC`. The event dispatcher would check this when the handler returns and call `Destroy()` if it is set. The `Destroy` method would set it to `NULL`.
I can live with the explicit `Destroy()` call as well. It should probably just find its way into the documentation as many people develop on one platform and users are using another.
Personally, I would probably prefer the `Destroy()` over the context manager as I dislike the additional indention level. The context manager is of course safer and more elegant.

But what I really don't understand, is why the destructor does not seem to get called right when the ref count reaches 0.
With Python 3.4 there were changes due to PEP442: https://docs.python.org/3/c-api/typeobj.html#Py_TPFLAGS_HAVE_FINALIZE
Could this be related?
But anyway, the matplotlib example shows that it's easy to get an additional reference and so one of the explicit versions is probably better than relying on the destructor.

**Comment by mesalu** [comment]
Things can take a deceptively long time to be collected. Especially if there is a cyclical reference involved somewhere. The collector categorizes objects into 'generations' based on how many sweeps they've survived. The longer they've stuck around, the longer it takes Python to get around to cleaning them up. 

See: Collector thresholds (from https://docs.python.org/3/library/gc.html)
>In order to decide when to run, the collector keeps track of the number object allocations and deallocations since the last collection. When the number of allocations minus the number of deallocations exceeds threshold0, collection starts. Initially only generation 0 is examined. If generation 0 has been examined more than threshold1 times since generation 1 has been examined, then generation 1 is examined as well. Similarly, threshold2 controls the number of collections of generation 1 before collecting generation 2.

I'm not sure how long this DC is sticking around, but if its alive long enough to be promoted to gen1 or 2, it'll take some time before being cleaned. It could be that by suspending the GC you were preventing it from being promoted and as such its collection was more timely upon exiting the handler.

Making changes to the event dispatcher is, frankly, out of the question. Phoenix wraps wxWidgets, with the intent of only making modifications to pythonize the underlying library. Tweaking the core event dispatcher like that is a slippery slope to maintenance hell.

If indentations really bother you, you can always add a method call to reduce indentation.

**Comment by DietmarSchwertberger** [comment]
OK, I think I've found the reason why the `PaintDC` did not get destroyed even when there was no reference left.

I have just created a Python proxy for the `PaintDC`. This proxy has a `__del__` method which just prints a message and destroys the original `PaintDC`.

The outcome was:
 * If the `PaintDC` instance `drawDC` just goes out of scope as the handler returns, then the `__del__` destructor is only called late, during the next gc run. If the gc is disabled, it will never be called.
 * When I add a `del drawDC` at the end of the handler, then the destructor is called immediately, if no additional reference is left. This even works with the gc being disabled completely. (With an unmodified Matplotlib, one reference could be left, though.)

So the actual behaviour of CPython seems to be different from 'common knowledge' and from what is written in the documentation. But anyway, the explicit calls of `Destroy()` or the context manager should always fix any issues.

With `del drawDC`:
```
Created   0x109a17240
Finishing 0x109a17240
__del__ 0x109a17240
Created   0x109a17240
Finishing 0x109a17240
__del__ 0x109a17240
Created   0x109a17240
Finishing 0x109a17240
```

Without:
```
Created   0x109a182b0
Finishing 0x109a182b0
Created   0x109a28b00
__del__ 0x109a182b0
Finishing 0x109a28b00
Created   0x105759a58
Finishing 0x105759a58
Created   0x109a43b00
Finishing 0x109a43b00
Created   0x1055ea7b8
Finishing 0x1055ea7b8
Created   0x109b39f98
Finishing 0x109b39f98
Created   0x109b7c390
__del__ 0x105759a58
__del__ 0x109a43b00
__del__ 0x1055ea7b8
__del__ 0x109b39f98
Finishing 0x109b7c390
```

The proxy:
```
class PaintDC:
    def __init__(self, window):
        self.__dict__['dc'] = wx.PaintDC(window)
    def __getattr__(self, name):
        return getattr(self.dc, name)
    def __setattr__(self, name, value):
        setattr(self.dc, name, value)
    def __del__(self):
        print("__del__", hex(id(self)))
        self.dc.Destroy()
        del self.dc
```

**Comment by RobinD42** [comment]
Thanks for the analysis @DietmarSchwertberger. I still have a suspicion that Matplotlib is doing something somewhere that is affecting things in some way, as this is the only case of this problem that has been seen so far. However I agree that calling `Destroy` at the end of the handler one way or another should be a safe solution.
